File: Problem3/AccountOverdrawDemo.java
package Problem3;


// Account Class
class Account {
    public String name;
    public int balance = 1000;
    // Constructor
    Account(String name, int balance) {
        this.name = name;
        this.balance = balance;
    }
    // Withdraw method
    public boolean withdraw(int amount,String person){
        try{
            Thread.sleep(100);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        if(balance >= amount){
            balance -= amount;
            System.out.println("Account Holder: "+person +"\nWithdrawal: " + amount + " | Current Balance: " + balance);
            return  true;
        } else {
            System.out.println("Account Holder: "+person+" \nInsufficient Balance");
        }
        return false;
    }
}


class AccountOverdrawDemo extends Thread{
    Account account;
    String person;
    public AccountOverdrawDemo(Account account,String person){
        this.account = account;
        this.person = person;
    }
    // Overriding the run method
    @Override
    public void run() {
        for(int i=0;i<5;i++){
            // Withdrawing 200 from the account
            account.withdraw(200,person);
        }
    }
    public static void main(String[] args) {
        // Creating a new Account
        Account acc1 = new Account("Johns and Sons",1000);
        // Creating two threads
        AccountOverdrawDemo john = new AccountOverdrawDemo(acc1,"John");
        AccountOverdrawDemo son = new AccountOverdrawDemo(acc1,"Junior John");
        //
        john.start();
        son.start();
    }
}

File: Problem3/AccountOverdrawDemo.js
//creating account class
class Account {
    constructor(name1, name2, balance) {
        this.name1 = name1;
        this.name2 = name2;
        this.balance = balance;
    }

    // method to withdraw amount from account
    async withdraw(amount, name) {
        //only if amount is less then amount will withdraw
        if (this.balance >= amount) {
            console.log(`${name} is withdrawing ${amount}`);
            await new Promise(resolve => setTimeout(resolve, 1000));
            this.balance -= amount;
            console.log(`New balance: ${this.balance}`);// withdraw succesful
            return true;
        } else {
            //if they don't have balance
            console.log(`${name} you do not have ${amount}`);
            return false;
        }
    }
}

class AccountOverdrawDemo {
    //run method to run tasks or threads 
    static async run() {
        //joint account holder's, account creation 
        const account = new Account('Ishaan', 'Akar', 1000);

        // Withdrawal 1
        const thread1 = async (name) => {
            while (true) {
                //randomly generating withdraw amount
                const flag = await account.withdraw(Math.floor(Math.random() * 300) + 100, name);//calling method
                if (!flag) break;
            }
        };

        // Withdrawal 2 
        const thread2 = async (name) => {
            while (true) {
                //randomly generating withdraw amount
                const flag = await account.withdraw(Math.floor(Math.random() * 500) + 100, name);//calling method
                if (!flag) break;
            }
        };

        await Promise.all([thread1('Ishaan'), thread2('Akar')]);
        console.log(`Final balance: ${account.balance}`);
    }
}

// Test AccountOverdrawDemo
console.log('Running AccountOverdrawDemo...\n');
AccountOverdrawDemo.run().catch(console.error);
